src/targets/TARGET_STM/USBHALHost_STM.cpp

```diff
@@ -91,7 +91,122 @@ uint32_t HAL_HCD_HC_GetType(HCD_HandleTypeDef *hhcd, uint8_t chnum)
 //  - URB_NOTREADY = a NAK, NYET, or not more than a couple of repeats of some of the errors that will
 //                   become URB_ERROR if they repeat several times in a row
 //
-#if ARC_USB_FULL_SIZE
+#if ARC_FS_OVER_HS
+//
+// In the underlying HAL code there are the following issues when running FS using the HS interface:
+// This effects the GIGA as it has no external PHY which is required in order to run at HS.
+//
+// 1. Transmitting length larger then packetsize can cause future issues reading, the NAK nightmare.
+// 2. Receiving multiple seperate packets can lead to lost data and the need to retry.
+//
+// So for transmitting we split the data into sizes of <= packet size and individually transfer them, only calling pPriv->transferCompleted() when all data is sent.
+// For receiving we receive all the data and call pPriv->transferCompleted().
+//
+// If we get NAKS on the control OUT EPs we make sure that USB_OTG_HCCHAR_CHDIS ans USB_OTG_HCCHAR_CHENA are reset to the correct values.
+//
+void HAL_HCD_HC_NotifyURBChange_Callback(HCD_HandleTypeDef *pHcd, uint8_t uChannel, HCD_URBStateTypeDef urbState)
+{
+  USBHALHost_Private_t *pPriv = (USBHALHost_Private_t *)(pHcd->pData);
+
+  HCTD *pTransferDescriptor = (HCTD *)pPriv->addr[uChannel];
+
+
+  if (pTransferDescriptor) 
+  {
+    uint32_t endpointType = pHcd->hc[uChannel].ep_type;
+
+    if ((endpointType == EP_TYPE_INTR)) 
+    {
+      // Disable the channel interrupt and retransfer below
+      // TOD: really this retransfer should be queued up and transfered on the SOF interupt based on interval from descriptor.
+      pTransferDescriptor->state = USB_TYPE_IDLE ;
+      HAL_HCD_DisableInt(pHcd, uChannel);
+      pTransferDescriptor->currBufPtr += HAL_HCD_HC_GetXferCount(pHcd, uChannel);
+    } 
+    else if ((endpointType == EP_TYPE_BULK) || (endpointType == EP_TYPE_CTRL)) 
+    {
+      uint8_t  uEpIsInput  = pHcd->hc[uChannel].ep_is_in;
+
+      switch(urbState)
+      {
+        case URB_NOTREADY:
+        {
+          if(!uEpIsInput)
+          {
+#if ARC_NO_RESUBMITREQUEST
+            // make sure that USB_OTG_HCCHAR_CHDIS and USB_OTG_HCCHAR_CHENA are reset to the correct values
+						uint32_t tmpreg;
+						tmpreg = USBx_HC(uChannel)->HCCHAR;
+						tmpreg &= ~USB_OTG_HCCHAR_CHDIS;
+						tmpreg |= USB_OTG_HCCHAR_CHENA;
+						USBx_HC(uChannel)->HCCHAR = tmpreg;
```

Comment:
```suggestion
            uint32_t tmpreg;
            tmpreg = USBx_HC(uChannel)->HCCHAR;
            tmpreg &= ~USB_OTG_HCCHAR_CHDIS;
            tmpreg |= USB_OTG_HCCHAR_CHENA;
            USBx_HC(uChannel)->HCCHAR = tmpreg;
```

Use spaces instead of tabs for indentation, as is done throughout the rest of the file.